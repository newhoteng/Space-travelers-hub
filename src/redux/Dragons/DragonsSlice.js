import { createSlice, createAsyncThunk } from '@reduxjs/toolkit';
import axios from 'axios';

const url = 'https://api.spacexdata.com/v4/dragons';

const initialState = {
  dragon: [],
  isLoading: false,
};

export const getDragons = createAsyncThunk('dragons/fetchDragons', async () => {
  const response = await axios.get(url);
  const dragon = response.data;
  return dragon.map((dragon) => ({
    id: dragon.id,
    name: dragon.name,
    description: dragon.description,
    image: dragon.flickr_images[0],
  }
  ));
});

const dragonSlice = createSlice({
  name: 'dragon',
  initialState,
  reducers: {
    reserveDragon: (state, action) => {
      const dragon = state.dragon.map((dragon) => {
        if (dragon.id === action.payload) return { ...dragon, reserved: !dragon.reserved };
        return dragon;
      });
      return { ...state, dragon };
    },
  },
  extraReducers: (builder) => {
    builder
      .addCase(getDragons.pending, (state) => ({
        ...state,
        isLoading: true,
      }))
      .addCase(getDragons.fulfilled, (state, action) => ({
        ...state,
        dragon: action.payload,
      }));
  },
});

export const { reserveDragon } = dragonSlice.actions;
export default dragonSlice.reducer;
