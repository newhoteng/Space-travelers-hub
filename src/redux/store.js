import { configureStore } from '@reduxjs/toolkit';
import missionsReducer from './missions/missionsSlice';
import dragonReducer from './Dragons/DragonsSlice';
import rocketReducer from './Rockets/RocketsSlice';

export const store = configureStore({
  reducer: {
    missions: missionsReducer,
    dragon: dragonReducer,
    rocket: rocketReducer,
  },
});

export default store;
