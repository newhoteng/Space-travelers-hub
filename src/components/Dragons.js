import { useEffect } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { getDragons, reserveDragon } from '../redux/Dragons/DragonsSlice';
import styles from '../styles/Dragon.module.css';

const Dragons = () => {
  const dispatch = useDispatch();
  const { dragon } = useSelector((store) => store.dragon);
  const { isLoading } = useSelector((state) => state.dragon);

  useEffect(() => {
    if (isLoading === false) dispatch(getDragons());
  }, [dispatch, isLoading]);

  return (
    <div className={styles.dragonsContainer}>
      {isLoading
        && dragon.map((dragon) => (
          <div className={styles.container} key={dragon.id}>
            <div className="article">
              <img
                src={dragon.image}
                alt={dragon.name}
                className="dragon-image"
              />
            </div>
            <div className="desc">
              <h2 className="dragon-title">{dragon.name}</h2>
              <p className={styles.dragonDetails}>
                {dragon.reserved && <span className="reserved">Reserved</span>}
                {dragon.description}
              </p>
              <button
                type="button"
                className={dragon.reserved ? styles.cancel : styles.reserve}
                onClick={() => dispatch(reserveDragon(dragon.id))}
              >
                {dragon.reserved ? 'Cancel Reservation' : 'Reserve Dragon'}
              </button>
            </div>
          </div>
        ))}
    </div>
  );
};

export default Dragons;
