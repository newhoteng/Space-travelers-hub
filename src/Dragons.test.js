import { render } from '@testing-library/react';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import { store } from './redux/store';

// Component imports
import Dragons from './components/Dragons';

import reducer, { reserveDragon } from './redux/Dragons/DragonsSlice';

describe('Dragons Slice', () => {
  test('should handle reserving a dragon', () => {
    const state = {
      dragon: [
        {
          id: '0', name: 'Dragon 1', description: 'The best', image: 'linkto image',
        },
        {
          id: '1', name: 'Dragon 2', description: 'Second best', image: 'linkto image',
        },
      ],
      status: false,
      error: null,
    };

    expect(reducer(state, reserveDragon('1'))).toEqual(
      {
        dragon: [
          {
            id: '0', name: 'Dragon 1', description: 'The best', image: 'linkto image',
          },
          {
            id: '1', name: 'Dragon 2', description: 'Second best', image: 'linkto image', reserved: true,
          },
        ],
        status: false,
        error: null,
      },
    );
  });
});

it('dragons renders correctly', () => {
  const dragons = renderer
    .create(
      <Provider store={store}>
        <Dragons page="/dragons">Dragons Page</Dragons>
      </Provider>,
    )
    .toJSON();
  expect(dragons).toMatchSnapshot();
});

it('renders Dragons component', () => {
  render(
    <Provider store={store}>
      <Dragons />
    </Provider>,
  );
});
