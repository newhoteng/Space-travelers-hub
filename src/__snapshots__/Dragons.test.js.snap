// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dragons renders correctly 1`] = `
<div
  className="dragonsContainer"
/>
`;
